File: src/_05_null_safety/NullSafety.java
package _05_null_safety;

public class NullSafety {
    public void nullSafetyString() {
        String value = "initial value";
        value = null;
    }

    public void nullSafetyClass() {
        Name name = new Name();
        name = null;
    }

    public void safetyCall(final Person person) {
        int result = 0;
        if (person != null && person.getName() != null && person.getName().getValue() != null) {
            result = person.getName().getValue().length();
        }
        System.out.println(result);
    }

    public void notNullAssert(final Person person) {
        if (person == null || person.getName() == null || person.getName().getValue() == null) {
            throw new NullPointerException();
        }

        int result = person.getName().getValue().length();
        System.out.println(result);
    }

    class Person {
        private Name name;

        public Name getName() {
            return name;
        }

        public void setName(Name name) {
            this.name = name;
        }
    }
    
    class Name {
        private String value;

        public String getValue() {
            return value;
        }

        public void setValue(String value) {
            this.value = value;
        }
    }
}

File: src/_05_null_safety/NullSafety.kt
package _05_null_safety

class Person(val name: Name?)
class Name(val value: String?)


fun nullSafetyString() {
    var value : String = "initial value"
    // value = null // compilation error
    
    var source : String? = "initial value"
    source = null
}

fun nullSafetyClass() {
    var name : Name = Name("")
    //name = null // compilation error
}


fun safeCalls(person : Person?) {
    val result = person?.name?.value?.length
    print(result)
}

fun notNullAssertionOperator(age: Name?) {
    val result = age!!.value!!.length
    print(result)
}

fun elvisOperator(value: String?) {
    val result = value ?: "default"
    val result2 = if (value == null)  "default" else value
    print(result)
}
